File: python/vector_class.py
import math

class Vector:
    def __init__(self, *args):
        if len(args) == 1 and isinstance(args[0], (list, tuple)):
            self.x, self.y, self.z = args[0]
        else:
            self.x, self.y, self.z = args

    def __add__(self, other):
        return Vector(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other):
        return Vector(self.x - other.x, self.y - other.y, self.z - other.z)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y and self.z == other.z

    def cross(self, other):
        return Vector(
            self.y * other.z - self.z * other.y,
            self.z * other.x - self.x * other.z,
            self.x * other.y - self.y * other.x
        )

    def dot(self, other):
        return self.x * other.x + self.y * other.y + self.z * other.z

    def to_a(self):
        return [self.x, self.y, self.z]

    def to_tuple(self):  # Added this method
        return (self.x, self.y, self.z)

    def __str__(self):
        return f"<{self.x}, {self.y}, {self.z}>"

    @property
    def magnitude(self):
        return math.sqrt(self.x ** 2 + self.y ** 2 + self.z ** 2)

File: ruby/vector_class.rb
# https://www.codewars.com/kata/532a69ee484b0e27120000b6/solutions/python

class Vector
  attr_reader :x, :y, :z

  def initialize(*args)
    if args.length == 1 && args.first.is_a?(Array)
      @x, @y, @z = args.first
    else
      @x, @y, @z = args
    end
  end

  def +(other)
    Vector.new(@x + other.x, @y + other.y, @z + other.z)
  end

  def -(other)
    Vector.new(@x - other.x, @y - other.y, @z - other.z)
  end

  def ==(other)
    @x == other.x && @y == other.y && @z == other.z
  end

  def cross(other)
    Vector.new(
      @y * other.z - @z * other.y,
      @z * other.x - @x * other.z,
      @x * other.y - @y * other.x
    )
  end

  def dot(other)
    @x * other.x + @y * other.y + @z * other.z
  end

  def to_a
    [@x, @y, @z]
  end

  def to_s
    "<#{@x}, #{@y}, #{@z}>"
  end

  def magnitude
    Math.sqrt(@x**2 + @y**2 + @z**2)
  end
end